src/Common/tests/System/Net/Http/HPackEncoder.cs

```diff
@@ -0,0 +1,194 @@
+
+        private static int EncodeString(string value, Span<byte> headerBlock, bool huffmanEncode)
+        {
+            byte[] data = Encoding.ASCII.GetBytes(value);
+            byte prefix;
+
+            if (!huffmanEncode)
+            {
+                prefix = 0;
+            }
+            else
+            {
+                int len = HuffmanEncoder.GetEncodedLength(data);
+
+                byte[] huffmanData = new byte[len];
+                HuffmanEncoder.Encode(data, huffmanData);
+
+                data = huffmanData;
+                prefix = 0x80;
+            }
+
+            int bytesGenerated = 0;
+
+            bytesGenerated += EncodeInteger(data.Length, prefix, 0x80, headerBlock);
+
+            data.AsSpan().CopyTo(headerBlock.Slice(bytesGenerated));
+            bytesGenerated += data.Length;
+
+            return bytesGenerated;
+        }
+
+        public static int EncodeInteger(int value, byte prefix, byte prefixMask, Span<byte> headerBlock)
+        {
+            byte prefixLimit = (byte)(~prefixMask);
+
+            if (value < prefixLimit)
+            {
+                headerBlock[0] = (byte)(prefix | value);
+                return 1;
+            }
+
+            headerBlock[0] = (byte)(prefix | prefixLimit);
+            int bytesGenerated = 1;
+
+            value -= prefixLimit;
+
+            while (value >= 0x80)
+            {
+                headerBlock[bytesGenerated] = (byte)((value & 0x7F) | 0x80);
+                value >>= 7;
+                bytesGenerated++;
+            }
+
+            headerBlock[bytesGenerated] = (byte)value;
+            bytesGenerated++;
+
+            return bytesGenerated;
+        }
+    }
+
+    public enum HPackFlags
+    {
+        /// <summary>
+        /// Encodes a header literal without indexing and without huffman encoding.
+        /// </summary>
+        None = 0,
+
+        /// <summary>
+        /// Applies Huffman encoding to the header's name.
+        /// </summary>
+        HuffmanEncodeName = 1,
+
+        /// <summary>
+        /// Applies Huffman encoding to the header's value.
+        /// </summary>
+        HuffmanEncodeValue = 2,
+
+        /// <summary>
+        /// Applies Huffman encoding to both the name and the value of the header.
+        /// </summary>
+        HuffmanEncode = HuffmanEncodeName | HuffmanEncodeValue,
+
+        /// <summary>
+        /// Encode a literal value without adding a new dynamic index. Intermediaries (such as a proxy) are still allowed to index the value when forwarding the header.
+        /// </summary>
+        WithoutIndexing = 0,
+
+        /// <summary>
+        /// Encode a literal value to a new dynamic index.
+        /// </summary>
+        NewIndexed = 4,
+
+        /// <summary>
+        /// Encode a literal value without adding a new dynamic index. Intermediaries (such as a proxy) must not index the value when forwarding the header.
+        /// </summary>
+        NeverIndexed = 8
+    }
+}
```

Comment:
Is this a second implementation of hpack encoding, separate from the product implementation?

Reply:
Hmm, I see, you consolidated existing test code, rather than writing it from scratch as part of this PR. Ok.